libraries/libfc/src/variant_object.cpp

```diff
@@ -302,6 +307,12 @@ namespace fc
       return *this;
    }
 
+   mutable_variant_object& mutable_variant_object::operator=( variant_object&& obj )
+   {
+      *_key_value = std::move(*obj._key_value);
+      return *this;
```

Comment:
```suggestion
      if (obj._key_value.use_count() == 1)
          *_key_value = std::move(*obj._key_value);
      else
          *_key_value = *obj._key_value;
      return *this;
```

Reply:
I thought about doing that. It is not thread safe which is why I held off. Restricting `variant_object` to use within single threads is probably ok. Another approach would be to remove the use of `shared_ptr` in `variant_object` but that is obviously a much bigger change.

So do you think we should just comment that `variant_object` use across threads is not thread safe and add this?

Reply:
I think thread safe, but approximate, and you are right there is a race condition. The doc says:
```
In multithreaded environment, the value returned by use_count is approximate (typical implementations use a [memory_order_relaxed](https://en.cppreference.com/w/cpp/atomic/memory_order) load)
```

So you are right if another thread copies the `variant_object` just after we read the use_count, we would incorrectly steal from the copy.

So really checking the `use_count` may be worse, because bad code would seem to work, only to fail on rare occasions. If we do use some variant_objects across threads, then my suggestion is incorrect I think.

In that case, maybe it is better to just remove these move constructor/assignment operators if we cannot be sure they are safe.

Reply:
I think we could `assert(use_count() == 1)`. The constructor version of this is `explicit` so my thought was that if you are moving into this constructor then you should only do so if you know you have exclusive ownership of the `variant_object`.  Maybe we should add a `copy_and_destroy( variant_object&& obj )` method so that is very explicit.

Reply:
I vote for adding the `assert()`, and also using my version above to make it safer in release mode just in case.

Reply:
Added assert()